executor/adapter.go

```diff
@@ -64,11 +64,47 @@ type recordSet struct {
 
 func (a *recordSet) Fields() []*ast.ResultField {
 	if len(a.fields) == 0 {
-		a.fields = schema2ResultFields(a.executor.Schema(), a.stmt.Ctx.GetSessionVars().CurrentDB)
+		if len(a.stmt.outputNames) > 0 {
+			a.fields = colNames2ResultFields(a.executor.Schema(), a.stmt.outputNames, a.stmt.Ctx.GetSessionVars().CurrentDB)
+		} else {
+			a.fields = schema2ResultFields(a.executor.Schema(), a.stmt.Ctx.GetSessionVars().CurrentDB)
+		}
 	}
 	return a.fields
 }
 
+func colNames2ResultFields(schema *expression.Schema, names []*expression.NamingForMySQLProtocol, defaultDB string) []*ast.ResultField {
+	rfs := make([]*ast.ResultField, 0, schema.Len())
+	for i := 0; i < schema.Len(); i++ {
+		dbName := names[i].DBName
+		if dbName.L == "" && names[i].TblName.L != "" {
+			dbName = model.NewCIStr(defaultDB)
+		}
+		origColName := names[i].OrigColName
+		if origColName.L == "" {
+			origColName = names[i].ColName
+		}
+		rf := &ast.ResultField{
+			Column:       &model.ColumnInfo{Name: origColName, FieldType: *schema.Columns[i].RetType},
+			ColumnAsName: names[i].ColName,
+			Table:        &model.TableInfo{Name: names[i].OrigTblName},
+			TableAsName:  names[i].TblName,
+			DBName:       dbName,
+		}
+		// This is for compatibility.
```

Comment:
Could you add more comments about this compatibility behavior? For example, you can post a link to the MySQL document.

Reply:
added